client/asset/btc/btc.go

```diff
@@ -5418,83 +5473,348 @@ func (btc *baseWallet) MaxFundingFees(numTrades uint32, feeRate uint64, options
+	}
+	if refundsSwap(msgTx) {
+		return asset.Refund, nil, nil, totalOut + fee, fee, nil
+	}
+
+	// Bond refunds
+	redeemsBond := func(msgTx *wire.MsgTx) (bool, *asset.BondTxInfo) {
+		var bondInfo *asset.BondTxInfo
+		isBond := func(segwit bool, contract []byte) bool {
+			const scriptVer = 0
+			lockTime, pkHash, err := dexbtc.ExtractBondDetailsV0(scriptVer, contract)
+			if err != nil {
+				return false
 			}
+			bondInfo = &asset.BondTxInfo{
+				AccountID: []byte{}, // Could look for the bond tx to get this, I guess.
+				LockTime:  uint64(lockTime),
+				BondID:    pkHash[:],
+			}
+			return true
 		}
+		return containsContractAtPushIndex(msgTx, 2, isBond), bondInfo
+	}
+	if isBondRedemption, bondInfo := redeemsBond(msgTx); isBondRedemption {
+		return asset.RedeemBond, nil, bondInfo, totalOut, fee, nil
 	}
 
+	allOutputsPayUs := func(msgTx *wire.MsgTx) bool {
+		for _, txOut := range msgTx.TxOut {
+			scriptClass, addrs, _, err := txscript.ExtractPkScriptAddrs(txOut.PkScript, btc.chainParams)
+			if err != nil {
+				btc.log.Errorf("ExtractPkScriptAddrs error: %w", err)
+				return false
+			}
+			switch scriptClass {
+			case txscript.PubKeyHashTy, txscript.WitnessV0PubKeyHashTy:
+			default:
+				return false
+			}
+			if len(addrs) != 1 { // sanity check
+				return false
+			}
+
+			addr := addrs[0]
+			owns, err := btc.node.ownsAddress(addr)
+			if err != nil {
+				btc.log.Errorf("ownsAddress error: %w", err)
+				return false
+			}
+			if !owns {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	if tx.Send && allOutputsPayUs(msgTx) {
+		if len(msgTx.TxOut) == 1 {
+			return asset.Acceleration, nil, nil, 0, fee, nil
+		}
+		return asset.Split, nil, nil, 0, fee, nil
+	}
+
+	txOutDirection := func(msgTx *wire.MsgTx) (in, out uint64) {
+		for _, txOut := range msgTx.TxOut {
+			_, addrs, _, err := txscript.ExtractPkScriptAddrs(txOut.PkScript, btc.chainParams)
+			if err != nil {
+				btc.log.Errorf("ExtractPkScriptAddrs error: %w", err)
+				continue
+			}
+
+			if len(addrs) == 0 { // sanity check
+				continue
+			}
+
+			addr := addrs[0]
+			owns, err := btc.node.ownsAddress(addr)
+			if err != nil {
+				btc.log.Errorf("ownsAddress error: %w", err)
+				continue
+			}
+			if owns {
+				in += uint64(txOut.Value)
+			} else {
+				out += uint64(txOut.Value)
+			}
+		}
+		return
+	}
+	in, out := txOutDirection(msgTx)
+
+	getRecipient := func(msgTx *wire.MsgTx, receive bool) *string {
+		for _, txOut := range msgTx.TxOut {
+			_, addrs, _, err := txscript.ExtractPkScriptAddrs(txOut.PkScript, btc.chainParams)
+			if err != nil {
+				btc.log.Errorf("ExtractPkScriptAddrs error: %w", err)
+				continue
+			}
+
+			if len(addrs) == 0 { // sanity check
+				continue
+			}
+
+			addr := addrs[0]
+			owns, err := btc.node.ownsAddress(addr)
+			if err != nil {
+				btc.log.Errorf("ownsAddress error: %w", err)
+				continue
+			}
+
+			if receive == owns {
+				str := addr.String()
+				return &str
+			}
```

Comment:
Why are we taking this precaution? Can't we just set it to the zeroth address regardless?

Reply:
i don't think you're reading this right. There's no precaution being taken. If this transaction is a sending transaction, we return an address not controlled by the wallet, otherwise we return an address that is controlled by the wallet.